test/distributed/_composable/fsdp/test_fully_shard_compile.py

```diff
@@ -0,0 +1,60 @@
+# Owner(s): ["oncall: distributed"]
+
+
+import torch
+from torch.distributed._composable.fsdp import fully_shard
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import FSDPTest, MLP
+from torch.testing._internal.common_utils import run_tests
+
+
+class TestFullyShardCompileCompute(FSDPTest):
+    @skip_if_lt_x_gpu(2)
+    def test_disable_compiling_hooks(self):
+        self.run_subtests(
+            {
+                "skip_fsdp_hooks": [False, True],
+            },
+            self._test_disable_compiling_hooks,
+        )
+
+    def _test_disable_compiling_hooks(
+        self,
+        skip_fsdp_hooks: bool,
+    ):
+        torch._dynamo.reset()
+        trace_rules_check_count = 0
+        HOOKS_FILE_NAME = "torch/distributed/_composable/fsdp/_fsdp_state.py"
+        HOOK_WRAPPER_NAME = "fsdp_hook_wrapper"
+
+        def patched_trace_rules_check(*args, **kwargs):
+            nonlocal trace_rules_check_count
+            f_code = args[0]
+            if (
+                hasattr(f_code, "co_filename")
+                and f_code.co_filename.endswith(HOOKS_FILE_NAME)
+                and f_code.co_name != HOOK_WRAPPER_NAME
+            ):
+                trace_rules_check_count += 1
+            return orig_trace_rules_check(*args, **kwargs)
+
+        original_skip_fsdp_hooks = torch._dynamo.config.skip_fsdp_hooks
+        orig_trace_rules_check = torch._dynamo.trace_rules.check
+        torch.distributed.barrier()
+        torch._dynamo.config.skip_fsdp_hooks = skip_fsdp_hooks
+        torch._dynamo.trace_rules.check = patched_trace_rules_check
+        model = MLP(4)
+        fully_shard(model)
+        model.compile()
+        model(torch.randn((4, 4), device="cuda"))
+        torch.distributed.barrier()
+        torch._dynamo.config.skip_fsdp_hooks = original_skip_fsdp_hooks
+        torch._dynamo.trace_rules.check = orig_trace_rules_check
+        if skip_fsdp_hooks:
+            self.assertEqual(trace_rules_check_count, 0)
+        else:
+            self.assertTrue(trace_rules_check_count > 0)
+
```

Comment:
FSDP hooks are in skipfiles already. we do counting in trace rules instead of using ``torch._dynamo.testing.CompileCounter``